neuraxle/checkpoints.py

```diff
@@ -19,185 +19,558 @@
+                current_id=current_id,
+                checkpoint_path=context.get_path()
+            )
+            data_inputs.append(checkpoint)
+        data_container.set_data_inputs(data_inputs)
+
+        if self.expected_output_checkpointer is None:
+            data_container.set_expected_outputs(None)
+            return data_container
+
+        expected_outputs = []
         for current_id in data_container.current_ids:
-            if not os.path.exists(self.get_checkpoint_file_path(current_id)):
+            checkpoint = self.expected_output_checkpointer.read(checkpoint_path=context.get_path(), current_id=current_id)
+            expected_outputs.append(checkpoint)
+        data_container.set_expected_outputs(expected_outputs)
+
+        return data_container
+
+    def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
+        """
+        Returns if the whole data container has been checkpointed.
+
+        :param data_container: data container to read checkpoint for
+        :type data_container: DataContainer
+        :param context: execution context to read checkpoint from
+        :type context: ExecutionContext
+        :return: data container checkpoint
+        :rtype: DataContainer
+        """
+        for current_id, data_input, _ in data_container:
+            if not self.data_input_checkpointer.checkpoint_exists(checkpoint_path=context.get_path(), current_id=current_id):
+                return False
+
+        if self.expected_output_checkpointer is None:
+            if context.get_execution_mode() == ExecutionMode.FIT or context.get_execution_mode() == ExecutionMode.FIT_TRANSFORM:
+                warnings.warn('Resuming Data Checkpoint {0} without an expected output checkpointer with Step Saver {1}. Perhaps, you wanted to have an expected output checkpointer ?'.format(context.get_path(), self.__class__.__name__))
+            return True
+
+        for current_id, _, expected_output in data_container:
+            if not self.expected_output_checkpointer.checkpoint_exists(checkpoint_path=context.get_path(), current_id=current_id):
                 return False
 
         return True
 
-    def get_checkpoint_file_path(self, current_id) -> str:
-        """
-        Returns the checkpoint file path for a data input id
 
-        :param current_id:
-        :return:
-        """
-        return os.path.join(
-            self.checkpoint_path,
-            '{0}.pickle'.format(current_id)
+class MiniCheckpoint(Checkpoint):
+    """
+    :class:`Checkpoint` with pickle mini data checkpointers wrapped in a :class:`MiniDataCheckpointerWrapper`, and the default step saving checkpointer.
+
+    .. seealso::
+        * :class:`Checkpoint`
+        * :class:`MiniDataCheckpointerWrapper`
+    """
+
+    def __init__(self):
+        Checkpoint.__init__(
+            self,
+            step_checkpointer=StepSavingCheckpointer(),
+            data_checkpointers=[
```

Comment:
Do we need to pass a list of data checkpointers? Is it still useful considering the Mini wrapper exists and can combine them for us in advance?

Reply:
HEY, I just rememberd again of an awesome idea I had in my fat notebook: To have a step that combines many barrier-like steps (like checkpointers, which are not data transformers). 

See my other PR comment above that starts with `This code is uselessly segregated` [...]. 

Big idea: this means the `Checkpointer` class might just be a pipeline which combines the checkpointer steps we pass to it, and it executes all the checkpoints one after another. This would simplify code and reuse existing Pipeline code. Because in the end, a Checkpoint is just something that chains checkpointers, and checkpointers are probably of type BaseSteps. Well we've just discovered that they are Base Steps if they were not already.